### PR TITLE
[Impeller] Special case color filter in tiled texture.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/119874

Currently there are two inefficiencies in the main wonderous carosel page:

1. We require two pass to tile the texture with opacity, then apply a color filter, these operations should be combined.
2. The final coverage is used for intermediate render passes, but tilling could be applied at the end.

Fixing these increases the FPS I see from ~90 up to 110. Getting closer!

I first attempt to generalize the passthrough code to handle tiled textures. In summary, this was a huge mess:  https://github.com/flutter/engine/compare/main...jonahwilliams:engine:uv_stuff?expand=1 

I tried to create a simpler implementation that takes advantage of the fact that the color filters we care about are almost always set on the same paint that is drawing the tiled texture. The tiled texture contents can be responsible for applying color filter in a subpass that is sized to the texture.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
TBD how we should test this, texture allocation or whatnot.

**Comment by jonahwilliams** [comment]
## before

before_patch


## after 

after_patch

File: impeller/renderer/allocator.cc
Comment (jonahwilliams) [inline impeller/renderer/allocator.cc:55]: Perhaps we should add/remove these with some ifdef.

Comment (jonahwilliams) [reply impeller/renderer/allocator.cc:55]: Another idea is that if we do this only in builds for engine unit tests, we could probably track a vector of allocated sizes which make assertions easier

File: impeller/renderer/allocator.h
Comment (chinmaygarde) [inline impeller/renderer/allocator.h:42]: I am not sure if you intend to [incorporate the allocator instrumentation into this patch](https://github.com/flutter/flutter/issues/120502#issuecomment-1426545766). If it helps speeds thing along for this patch, maybe we can back this out and land this patch while working on the instrumentation separately. Your call.

Comment (jonahwilliams) [reply impeller/renderer/allocator.h:42]: I think now that I have an idea of how we will test this, I feel more comfortable backing this change out and working on it separately

File: impeller/entity/contents/tiled_texture_contents.h
Comment (chinmaygarde) [inline impeller/entity/contents/tiled_texture_contents.h:50]: It'd be good to note why this this is here as a naive implementation (like the one earlier) shouldn't have to care about the filter.

**Review by chinmaygarde — APPROVED** [review]
LGTM after backing out the allocator instrumentation is backed out. We discussed that this is would be fixed in a separate patch.

File: impeller/aiks/paint.cc
Comment (bdero) [inline impeller/aiks/paint.cc:105]: Won't this will break if the filter is set after the color source?

Comment (bdero) [reply impeller/aiks/paint.cc:105]: Oh, it's capturing a ref to the paint.

Comment (jonahwilliams) [reply impeller/aiks/paint.cc:105]: We can lose this optimization if its making canvas weird. now that we're handling filters better in general its not as necessary.

Comment (jonahwilliams) [reply impeller/aiks/paint.cc:105]: We can lose this optimization if its making canvas weird. now that we're handling filters better in general its not as necessary.

Comment (bdero) [reply impeller/aiks/paint.cc:105]: Nah I think it's alright. Just had to rethink a refactor I'm working on.

**Pull request merged**
